Translator restart if disconnected from upstream

lorbax opened the pull request:
When the jdc changes upstream, it disconnects the TProxy.
This PR addresses this issue by making the TProxy reconnecting to the JDC.

Removed the part related to the MG test.

How to test this PR:
The idea is to have a pool that rejects a share. It is possible to do that by tweaking the part of the code needed to check the target, forcing the pool to calculate a wrong hash for a share. 
Following @GitGab19 [suggestion](https://github.com/stratum-mining/stratum/issues/844#issuecomment-2118864024), the part of the code responsible for that is  [`roles-logc-sv2::channel_logic::channel_factory::on_submit_shares_extended(...)`](https://github.com/stratum-mining/stratum/blob/dev/protocols/v2/roles-logic-sv2/src/channel_logic/channel_factory.rs#L1146). Since other roles check the share using the same function, directly modifying this would have the consequence that no share at all is ever sent to the pool. If I understood correctly, the way to fix this issue is to apply @GitGab19 suggestion to a clone of function `on_submit_shares_extended`, which can be called call `on_submit_shares_extended_pool`. After that, modify the implementation of [`handle_submit_share_extended`](https://github.com/stratum-mining/stratum/blob/dev/roles/pool/src/lib/mining_pool/message_handler.rs#L151) trait function  of `ParseDownstreamMiningMessages` trait in the following way
```rust
fn handle_submit_shares_extended(
        &mut self,
        m: SubmitSharesExtended,
    ) -> Result<SendTo<()>, Error> {
        let res = self
            .channel_factory
            //.safe_lock(|cf| cf.on_submit_shares_extended(m.clone()))
            .safe_lock(|cf| cf.on_submit_shares_extended_pool(m.clone()))
            .map_err(|e| roles_logic_sv2::Error::PoisonLock(e.to_string()))?;
        ...
    }
```
 - run a "good" pool on port 34254, this is the pool that the jdc will fallback to.
 - run a "bad" pool with the modification above on port 44254, this pool will refuse the first share submitted to the upstream (recall to adjust the channel hashrate in the tproxy config)
 - run a jdc client with these upstreams in the config
```toml
[[upstreams]]
authority_pubkey = "9auqWEzQDVyd2oe1JVGFLMLHZtCo2FFqZwtKA5gd9xbuEu7PH72"
pool_address = "127.0.0.1:44254"
jd_address = "127.0.0.1:34264"
pool_signature = "Stratum v2 SRI Pool"
[[upstreams]]
authority_pubkey = "9auqWEzQDVyd2oe1JVGFLMLHZtCo2FFqZwtKA5gd9xbuEu7PH72"
pool_address = "127.0.0.1:34254"
jd_address = "127.0.0.1:34264"
pool_signature = "Stratum v2 SRI Pool"
```
 - run the downstreams (usually tproxy + sv1 miner)

Shourya742 reviewed (changes requested):
Concept ACK
This approach can be optimized. Currently, we are spawning a new task immediately following the task we intend to terminate. A more efficient method would involve incorporating the receiver as one of the branches in the select! block within the task we wish to terminate. This adjustment would eliminate the need for an additional asynchronous task, thereby facilitating a more graceful shutdown rather than an abrupt one.
A possible code snippet for one such task can be: 
```
let socket_reader_task = tokio::task::spawn(async move {
    let reader = BufReader::new(&*socket_reader);
    let mut messages = FramedRead::new(
        async_compat::Compat::new(reader),
        LinesCodec::new_with_max_length(MAX_LINE_LENGTH),
    );

    loop {
        tokio::select! {
            res = messages.next().fuse() => {
                match res {
                    Some(Ok(incoming)) => {
                        debug!("Receiving from Mining Device {}: {:?}", &host_, &incoming);
                        let incoming: json_rpc::Message = handle_result!(tx_status_reader, serde_json::from_str(&incoming));

                        if let v1::Message::StandardRequest(standard_req) = incoming {
                            if let Ok(Submit{..}) = standard_req.try_into() {
                                handle_result!(tx_status_reader, Self::save_share(self_.clone()));
                            }
                        }

                        let res = Self::handle_incoming_sv1(self_.clone(), incoming).await;
                        handle_result!(tx_status_reader, res);
                    },
                    Some(Err(_)) => {
                        handle_result!(tx_status_reader, Err(Error::Sv1MessageTooLong));
                    },
                    None => {
                        handle_result!(tx_status_reader, Err(
                            std::io::Error::new(
                                std::io::ErrorKind::ConnectionAborted,
                                "Connection closed by client"
                            )
                        ));
                        break;                    
                      }
                }
            },
            _ = cancellation_token_mining_device.cancelled().fuse() => {
                warn!("Cancellation token triggered: Initiating shutdown of sv1 downstream reader");
                break; 
            },
            _ = rx_shutdown_clone.recv().fuse() => {
                warn!("Shutdown signal received: Initiating shutdown of sv1 downstream reader");
                break;
            }
        }
    }
    warn!("Downstream: Reader task is terminating. Performing cleanup.");
    kill(&tx_shutdown_clone).await; 
});
```

lorbax commented on roles/translator/src/main.rs:
```diff
@@ -77,6 +82,120 @@ async fn main() {
+}
+
+#[allow(clippy::too_many_arguments)]
+async fn start<'a>(
+    rx_sv2_submit_shares_ext: Receiver<SubmitSharesExtended<'static>>,
+    tx_sv2_submit_shares_ext: Sender<SubmitSharesExtended<'static>>,
+    tx_sv2_new_ext_mining_job: Sender<NewExtendedMiningJob<'static>>,
+    tx_sv2_set_new_prev_hash: Sender<SetNewPrevHash<'static>>,
+    tx_sv2_extranonce: Sender<(ExtendedExtranonce, u32)>,
+    rx_sv2_extranonce: Receiver<(ExtendedExtranonce, u32)>,
+    rx_sv2_set_new_prev_hash: Receiver<SetNewPrevHash<'static>>,
+    rx_sv2_new_ext_mining_job: Receiver<NewExtendedMiningJob<'static>>,
+    rx_sv1_downstream: Receiver<DownstreamMessages>,
+    tx_sv1_bridge: Sender<DownstreamMessages>,
+    tx_sv1_notify: broadcast::Sender<server_to_client::Notify<'static>>,
+    target: Arc<Mutex<Vec<u8>>>,
+    tx_status: async_channel::Sender<Status<'static>>,
+    cancellation_token: CancellationToken,
+) {
```
Yes but at the end you have to construct the struct with a `new()` method or some `add_data_1()`, `add_data_2()` ecc. but eventually that amount of data is gonna needed, one way or another. 
Am I missing something?

lorbax commented:
> Concept ACK This approach can be optimized. Currently, we are spawning a new task immediately following the task we intend to terminate. A more efficient method would involve incorporating the receiver as one of the branches in the select! block within the task we wish to terminate. This adjustment would eliminate the need for an additional asynchronous task, thereby facilitating a more graceful shutdown rather than an abrupt one. A possible code snippet for one such task can be:
> 
> ```
> let socket_reader_task = tokio::task::spawn(async move {
>     let reader = BufReader::new(&*socket_reader);
>     let mut messages = FramedRead::new(
>         async_compat::Compat::new(reader),
>         LinesCodec::new_with_max_length(MAX_LINE_LENGTH),
>     );
> 
>     loop {
>         tokio::select! {
>             res = messages.next().fuse() => {
>                 match res {
>                     Some(Ok(incoming)) => {
>                         debug!("Receiving from Mining Device {}: {:?}", &host_, &incoming);
>                         let incoming: json_rpc::Message = handle_result!(tx_status_reader, serde_json::from_str(&incoming));
> 
>                         if let v1::Message::StandardRequest(standard_req) = incoming {
>                             if let Ok(Submit{..}) = standard_req.try_into() {
>                                 handle_result!(tx_status_reader, Self::save_share(self_.clone()));
>                             }
>                         }
> 
>                         let res = Self::handle_incoming_sv1(self_.clone(), incoming).await;
>                         handle_result!(tx_status_reader, res);
>                     },
>                     Some(Err(_)) => {
>                         handle_result!(tx_status_reader, Err(Error::Sv1MessageTooLong));
>                     },
>                     None => {
>                         handle_result!(tx_status_reader, Err(
>                             std::io::Error::new(
>                                 std::io::ErrorKind::ConnectionAborted,
>                                 "Connection closed by client"
>                             )
>                         ));
>                         break;                    
>                       }
>                 }
>             },
>             _ = cancellation_token_mining_device.cancelled().fuse() => {
>                 warn!("Cancellation token triggered: Initiating shutdown of sv1 downstream reader");
>                 break; 
>             },
>             _ = rx_shutdown_clone.recv().fuse() => {
>                 warn!("Shutdown signal received: Initiating shutdown of sv1 downstream reader");
>                 break;
>             }
>         }
>     }
>     warn!("Downstream: Reader task is terminating. Performing cleanup.");
>     kill(&tx_shutdown_clone).await; 
> });
> ```

That is a particular task, but how would you do with this task? I could not find a better solution.
```rust
        let task = tokio::task::spawn(async move {
            loop {
                let msg = handle_result!(tx_status, rx_sv1_downstream.clone().recv().await);

                match msg {
                    DownstreamMessages::SubmitShares(share) => {
                        handle_result!(
                            tx_status,
                            Self::handle_submit_shares(self_.clone(), share).await
                        );
                    }
                    DownstreamMessages::SetDownstreamTarget(new_target) => {
                        handle_result!(
                            tx_status,
                            Self::handle_update_downstream_target(self_.clone(), new_target)
                        );
                    }
                };
            }
        });
        tokio::task::spawn(async move {
            tokio::select! {
                _ = cancellation_token_handle_downstream.cancelled() => {
                    task.abort();
                    warn!("Shutting down handle_result task");
                },
            }
        });
```
and what do you think about using a mutex as JoinaHandler` collector for spawned tasks and close all of them in task in the main?

lorbax commented on roles/translator/src/lib/downstream_sv1/downstream.rs:
```diff
@@ -205,15 +208,24 @@ impl Downstream {
             kill(&tx_shutdown_clone).await;
             warn!("Downstream: Shutting down sv1 downstream reader");
         });
+        tokio::task::spawn(async move {
+            tokio::select! {
+                _ = cancellation_token_mining_device.cancelled() => {
+                    socket_reader_task.abort();
+                    warn!("Shutting down sv1 downstream reader");
+                },
+            }
+        });
```
I apply this change to every task

Shourya742 commented:
> > Concept ACK This approach can be optimized. Currently, we are spawning a new task immediately following the task we intend to terminate. A more efficient method would involve incorporating the receiver as one of the branches in the select! block within the task we wish to terminate. This adjustment would eliminate the need for an additional asynchronous task, thereby facilitating a more graceful shutdown rather than an abrupt one. A possible code snippet for one such task can be:
> > ```
> > let socket_reader_task = tokio::task::spawn(async move {
> >     let reader = BufReader::new(&*socket_reader);
> >     let mut messages = FramedRead::new(
> >         async_compat::Compat::new(reader),
> >         LinesCodec::new_with_max_length(MAX_LINE_LENGTH),
> >     );
> > 
> >     loop {
> >         tokio::select! {
> >             res = messages.next().fuse() => {
> >                 match res {
> >                     Some(Ok(incoming)) => {
> >                         debug!("Receiving from Mining Device {}: {:?}", &host_, &incoming);
> >                         let incoming: json_rpc::Message = handle_result!(tx_status_reader, serde_json::from_str(&incoming));
> > 
> >                         if let v1::Message::StandardRequest(standard_req) = incoming {
> >                             if let Ok(Submit{..}) = standard_req.try_into() {
> >                                 handle_result!(tx_status_reader, Self::save_share(self_.clone()));
> >                             }
> >                         }
> > 
> >                         let res = Self::handle_incoming_sv1(self_.clone(), incoming).await;
> >                         handle_result!(tx_status_reader, res);
> >                     },
> >                     Some(Err(_)) => {
> >                         handle_result!(tx_status_reader, Err(Error::Sv1MessageTooLong));
> >                     },
> >                     None => {
> >                         handle_result!(tx_status_reader, Err(
> >                             std::io::Error::new(
> >                                 std::io::ErrorKind::ConnectionAborted,
> >                                 "Connection closed by client"
> >                             )
> >                         ));
> >                         break;                    
> >                       }
> >                 }
> >             },
> >             _ = cancellation_token_mining_device.cancelled().fuse() => {
> >                 warn!("Cancellation token triggered: Initiating shutdown of sv1 downstream reader");
> >                 break; 
> >             },
> >             _ = rx_shutdown_clone.recv().fuse() => {
> >                 warn!("Shutdown signal received: Initiating shutdown of sv1 downstream reader");
> >                 break;
> >             }
> >         }
> >     }
> >     warn!("Downstream: Reader task is terminating. Performing cleanup.");
> >     kill(&tx_shutdown_clone).await; 
> > });
> > ```
> 
> That is a particular task, but how would you do with this task? I could not find a better solution.
> 
> ```rust
>         let task = tokio::task::spawn(async move {
>             loop {
>                 let msg = handle_result!(tx_status, rx_sv1_downstream.clone().recv().await);
> 
>                 match msg {
>                     DownstreamMessages::SubmitShares(share) => {
>                         handle_result!(
>                             tx_status,
>                             Self::handle_submit_shares(self_.clone(), share).await
>                         );
>                     }
>                     DownstreamMessages::SetDownstreamTarget(new_target) => {
>                         handle_result!(
>                             tx_status,
>                             Self::handle_update_downstream_target(self_.clone(), new_target)
>                         );
>                     }
>                 };
>             }
>         });
>         tokio::task::spawn(async move {
>             tokio::select! {
>                 _ = cancellation_token_handle_downstream.cancelled() => {
>                     task.abort();
>                     warn!("Shutting down handle_result task");
>                 },
>             }
>         });
> ```
> 
> and what do you think about using a mutex as JoinaHandler` collector for spawned tasks and close all of them in task in the main?
Ah, I missed it! In that case, aggregating JoinHandles with simultaneous termination would make more sense.

lorbax commented:
Applied the changes suggested by @Shourya742

lorbax commented:
> > > Concept ACK This approach can be optimized. Currently, we are spawning a new task immediately following the task we intend to terminate. A more efficient method would involve incorporating the receiver as one of the branches in the select! block within the task we wish to terminate. This adjustment would eliminate the need for an additional asynchronous task, thereby facilitating a more graceful shutdown rather than an abrupt one. A possible code snippet for one such task can be:
> > > ```
> > > let socket_reader_task = tokio::task::spawn(async move {
> > >     let reader = BufReader::new(&*socket_reader);
> > >     let mut messages = FramedRead::new(
> > >         async_compat::Compat::new(reader),
> > >         LinesCodec::new_with_max_length(MAX_LINE_LENGTH),
> > >     );
> > > 
> > >     loop {
> > >         tokio::select! {
> > >             res = messages.next().fuse() => {
> > >                 match res {
> > >                     Some(Ok(incoming)) => {
> > >                         debug!("Receiving from Mining Device {}: {:?}", &host_, &incoming);
> > >                         let incoming: json_rpc::Message = handle_result!(tx_status_reader, serde_json::from_str(&incoming));
> > > 
> > >                         if let v1::Message::StandardRequest(standard_req) = incoming {
> > >                             if let Ok(Submit{..}) = standard_req.try_into() {
> > >                                 handle_result!(tx_status_reader, Self::save_share(self_.clone()));
> > >                             }
> > >                         }
> > > 
> > >                         let res = Self::handle_incoming_sv1(self_.clone(), incoming).await;
> > >                         handle_result!(tx_status_reader, res);
> > >                     },
> > >                     Some(Err(_)) => {
> > >                         handle_result!(tx_status_reader, Err(Error::Sv1MessageTooLong));
> > >                     },
> > >                     None => {
> > >                         handle_result!(tx_status_reader, Err(
> > >                             std::io::Error::new(
> > >                                 std::io::ErrorKind::ConnectionAborted,
> > >                                 "Connection closed by client"
> > >                             )
> > >                         ));
> > >                         break;                    
> > >                       }
> > >                 }
> > >             },
> > >             _ = cancellation_token_mining_device.cancelled().fuse() => {
> > >                 warn!("Cancellation token triggered: Initiating shutdown of sv1 downstream reader");
> > >                 break; 
> > >             },
> > >             _ = rx_shutdown_clone.recv().fuse() => {
> > >                 warn!("Shutdown signal received: Initiating shutdown of sv1 downstream reader");
> > >                 break;
> > >             }
> > >         }
> > >     }
> > >     warn!("Downstream: Reader task is terminating. Performing cleanup.");
> > >     kill(&tx_shutdown_clone).await; 
> > > });
> > > ```
> > 
> > 
> > That is a particular task, but how would you do with this task? I could not find a better solution.
> > ```rust
> >         let task = tokio::task::spawn(async move {
> >             loop {
> >                 let msg = handle_result!(tx_status, rx_sv1_downstream.clone().recv().await);
> > 
> >                 match msg {
> >                     DownstreamMessages::SubmitShares(share) => {
> >                         handle_result!(
> >                             tx_status,
> >                             Self::handle_submit_shares(self_.clone(), share).await
> >                         );
> >                     }
> >                     DownstreamMessages::SetDownstreamTarget(new_target) => {
> >                         handle_result!(
> >                             tx_status,
> >                             Self::handle_update_downstream_target(self_.clone(), new_target)
> >                         );
> >                     }
> >                 };
> >             }
> >         });
> >         tokio::task::spawn(async move {
> >             tokio::select! {
> >                 _ = cancellation_token_handle_downstream.cancelled() => {
> >                     task.abort();
> >                     warn!("Shutting down handle_result task");
> >                 },
> >             }
> >         });
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and what do you think about using a mutex as JoinaHandler` collector for spawned tasks and close all of them in task in the main?
> 
> Ah, I missed it! In that case, aggregating JoinHandles with simultaneous termination would make more sense.

We can discuss it at the call, when I realized using a mutex as a joinhandler collector for the tasks, I liked it but I do not want to change it if not necessary.

plebhash commented on test/message-generator/test/change-upstream/change-upstream.json:
I'm confused. I don't see MG starting any role here, which usually I see on the `setup_commands` field.

Is this testing tProxy? if so, how? What am I missing?

Also, there's no shell script for invoking this MG test, and it would also be good to add it to CI.

lorbax replied:
To replicate the issue and to test the fix is info contained in the messages of the two commits.

GitGab19 replied:
@plebhash I think that @lorbax is referring this: https://github.com/stratum-mining/stratum/pull/1001/commits/c54cb76cd7c92ad7444d3a4092cc823a5bf75fe8

lorbax replied:
My idea is to include a MG test in the CI, testing both this scenario and when there is solo fallback (which should be ready soon). Also, would be good to add a new functionality to the MG, namely the possibility to ignore all the messages until one of particular type is received. 
This is because at the same moment the translator can send a SetCustomMiningJob, an UpdateChannel and a SubmitShare to the upstream. Right now there is no way to this "hey ignore the first two message types and send back an error whenever you receive a SubmitShare".

plebhash replied:
> My idea is to include a MG test in the CI, testing both this scenario and when there is solo fallback (which should be ready soon). Also, would be good to add a new functionality to the MG, namely the possibility to ignore all the messages until one of particular type is received.

is that going to be done on a separate PR?

lorbax replied:
I think that it is a good idea. This PR has priority in my opinion. We can either wait to have the proper feature on the MG side, for me is the same.

GitGab19 commented:
@lorbax is it ready for a final review?
Did you complete the changes we discussed last week during dev call?

lorbax commented:
> @lorbax is it ready for a final review? Did you complete the changes we discussed last week during dev call?

yes!

Shourya742 commented on roles/translator/src/lib/downstream_sv1/downstream.rs:
```diff
@@ -242,11 +251,18 @@ impl Downstream {
             );
         });
+        let _ = task_collector_new_sv1_message_no_transl.safe_lock(|a| {
+            a.push((
+                socket_writer_task.abort_handle(),
+                "socket_writer_task".to_string(),
+            ))
```
Any specific reason adding context to abort handle?

plebhash replied:
it seems the purpose of these strings is so that `kill_tasks` can print them to the console, like this:
```
2024-08-01T23:36:40.372202Z  WARN translator_sv2: Killed task: "notify_task"
2024-08-01T23:36:40.372229Z  WARN translator_sv2: Killed task: "socket_writer_task"
2024-08-01T23:36:40.372233Z  WARN translator_sv2: Killed task: "socket_reader_task"
2024-08-01T23:36:40.372237Z  WARN translator_sv2: Killed task: "accept_connections"
2024-08-01T23:36:40.372240Z  WARN translator_sv2: Killed task: "handle_downstream_message"
2024-08-01T23:36:40.372244Z  WARN translator_sv2: Killed task: "handle_new_extended_mining_job"
2024-08-01T23:36:40.372247Z  WARN translator_sv2: Killed task: "handle_new_prev_hash"
2024-08-01T23:36:40.372251Z  WARN translator_sv2: Killed task: "handle_submit"
2024-08-01T23:36:40.372258Z  WARN translator_sv2: Killed task: "parse_incoming"
2024-08-01T23:36:40.372360Z  WARN translator_sv2: Killed task: "start_diff_management"
2024-08-01T23:36:40.372365Z  WARN translator_sv2: Killed task: "init task"
```

plebhash commented on roles/translator/config-examples/tproxy-config-local-jdc-example.toml:
it looks like this file was committed by mistake

plebhash commented:
yesterday we had a PR review club, where we dived into #1001 (which aims to fix #1004)

we found some issues (`UpstreamShutdown(UpstreamIncoming(UnexpectedMessage(1)))`), and after further investigation I reported it into #1079 and fixed via #1080

after that, I was able to manually reproduce #1004, but only with the trick described by @GitGab19 on this comment: https://github.com/stratum-mining/stratum/issues/844#issuecomment-2118864024

basically I launched one pool with this patch (listening on port 44254), and another pool without any patches (listening on port 34254)

then I modified JDC config so it would have two `[[upstreams]]` sections:
```
[[upstreams]]
authority_pubkey = "9auqWEzQDVyd2oe1JVGFLMLHZtCo2FFqZwtKA5gd9xbuEu7PH72"
pool_address = "127.0.0.1:44254"
jd_address = "127.0.0.1:34264"
# Pool signature (string to be included in coinbase tx)
pool_signature = "Stratum v2 SRI Pool"

[[upstreams]]
authority_pubkey = "9auqWEzQDVyd2oe1JVGFLMLHZtCo2FFqZwtKA5gd9xbuEu7PH72"
pool_address = "127.0.0.1:34254"
jd_address = "127.0.0.1:34264"
# Pool signature (string to be included in coinbase tx)
pool_signature = "Stratum v2 SRI Pool"
```

finally, I was able to verify that this PR is allowing tProxy to restart after the patched pool rejects a valid share

---

unfortunately the MG mock for a bad pool isn't working reliably... I tried following the instructions from 3d8793a3fc5aa087737dc3d207a6b54cf68858f0 multiple times but the behavior isn't deterministic

given that:
- #1059 was de-prioritized
- the instructions on 3d8793a3fc5aa087737dc3d207a6b54cf68858f0 are already highly manual, without deterministic automation
- the manual instructions above are sufficient for reproducing

I think we can probably drop 3d8793a3fc5aa087737dc3d207a6b54cf68858f0 and only use this PR to fix the bug

lorbax commented:
> yesterday we had a PR review club, where we dived into this PR
> 
> we found some issues (`UpstreamShutdown(UpstreamIncoming(UnexpectedMessage(1)))`), and after further investigation I reported it into #1079 and fixed via #1080
> 
> after that, I was able to manually reproduce #1004, but only with the trick described by @GitGab19 here: [#844 (comment)](https://github.com/stratum-mining/stratum/issues/844#issuecomment-2118864024)
> 
> basically I launched one pool with this patch (listening on port 44254), and another pool without any patches (listening on port 34254)
> 
> then I modified JDC config so it would have two `[[upstreams]]` sections:
> 
> ```
> [[upstreams]]
> authority_pubkey = "9auqWEzQDVyd2oe1JVGFLMLHZtCo2FFqZwtKA5gd9xbuEu7PH72"
> pool_address = "127.0.0.1:44254"
> jd_address = "127.0.0.1:34264"
> # Pool signature (string to be included in coinbase tx)
> pool_signature = "Stratum v2 SRI Pool"
> 
> [[upstreams]]
> authority_pubkey = "9auqWEzQDVyd2oe1JVGFLMLHZtCo2FFqZwtKA5gd9xbuEu7PH72"
> pool_address = "127.0.0.1:34254"
> jd_address = "127.0.0.1:34264"
> # Pool signature (string to be included in coinbase tx)
> pool_signature = "Stratum v2 SRI Pool"
> ```
> 

I already had included a jdc config for testing the PR
[https://github.com/lorbax/stratum/blob/translator-restart-if-disconnected-from-upstream/test/config/change-upstream/jdc-config-local-example-change-upstream.toml](https://github.com/lorbax/stratum/blob/translator-restart-if-disconnected-from-upstream/test/config/change-upstream/jdc-config-local-example-change-upstream.toml)
so that you do have to modify it. It is described in the PR that all the configs to be used are in `test/config/change-upstream/`

> unfortunately the MG mock for a bad pool isn't working reliably... I tried following the instructions from [3d8793a](https://github.com/stratum-mining/stratum/commit/3d8793a3fc5aa087737dc3d207a6b54cf68858f0) multiple times but the behavior isn't deterministic

The only thing is that the MG should need a new [MG feature](https://github.com/stratum-mining/stratum/issues/1059) in order to work flawlessly, but I used the MG test yesterday to replicate the issue and to test the fix (also after the review club) and it does the job. Can you provide more information about the difficulties or inconsistencies that you got using that test? "isn't working reliably" is a bit vague.

plebhash commented:
apologies for the lack of detail

I started all roles by following the instructions on 3d8793a3fc5aa087737dc3d207a6b54cf68858f0 and using all config files under `test/config/change-upstream`

the mock pool is killed immediately after the `SubmitShares.Error` is triggered, and as a consequence JDC does not try to fallback, it just gets stuck on trying to reconnect to the first pool

here's the JDC logs:
```
2024-08-02T15:33:49.535566Z  INFO jd_client::lib::upstream_sv2::upstream: Set custom mining job success 1
2024-08-02T15:33:58.853289Z  INFO roles_logic_sv2::handlers::mining: Received UpdateChannel->Extended message
2024-08-02T15:34:19.304315Z ERROR roles_logic_sv2::handlers::mining: Received SubmitSharesError with error code invalid-channel-id
2024-08-02T15:34:19.307149Z ERROR network_helpers_sv2::noise_connection_tokio: Disconnected from client while reading : early eof - 127.0.0.1:44254
2024-08-02T15:34:19.307180Z ERROR jd_client::lib::status: Error: ChannelErrorReceiver(RecvError)
2024-08-02T15:34:19.307225Z ERROR jd_client::lib: SHUTDOWN from: Channel receive error: `RecvError`
2024-08-02T15:34:20.319042Z ERROR network_helpers_sv2::noise_connection_tokio: Disconnecting from client due to error receiving: receiving from an empty and closed channel - 127.0.0.1:8442
2024-08-02T15:34:20.321992Z ERROR network_helpers_sv2::noise_connection_tokio: Disconnected from client while reading : early eof - 127.0.0.1:8442
2024-08-02T15:34:21.321514Z ERROR jd_client::lib::upstream_sv2::upstream: Failed to connect to Upstream role at 127.0.0.1:44254, retrying in 5s: Connection refused (os error 61)
2024-08-02T15:34:26.331207Z ERROR jd_client::lib::upstream_sv2::upstream: Failed to connect to Upstream role at 127.0.0.1:44254, retrying in 5s: Connection refused (os error 61)
2024-08-02T15:34:31.335578Z ERROR jd_client::lib::upstream_sv2::upstream: Failed to connect to Upstream role at 127.0.0.1:44254, retrying in 5s: Connection refused (os error 61)
2024-08-02T15:34:36.341290Z ERROR jd_client::lib::upstream_sv2::upstream: Failed to connect to Upstream role at 127.0.0.1:44254, retrying in 5s: Connection refused (os error 61)
2024-08-02T15:34:41.343409Z ERROR jd_client::lib::upstream_sv2::upstream: Failed to connect to Upstream role at 127.0.0.1:44254, retrying in 5s: Connection refused (os error 61)
2024-08-02T15:34:46.346232Z ERROR jd_client::lib::upstream_sv2::upstream: Failed to connect to Upstream role at 127.0.0.1:44254, retrying in 5s: Connection refused (os error 61)
2024-08-02T15:34:51.348015Z ERROR jd_client::lib::upstream_sv2::upstream: Failed to connect to Upstream role at 127.0.0.1:44254, retrying in 5s: Connection refused (os error 61)
2024-08-02T15:34:56.353763Z ERROR jd_client::lib::upstream_sv2::upstream: Failed to connect to Upstream role at 127.0.0.1:44254, retrying in 5s: Connection refused (os error 61)
2024-08-02T15:35:01.357632Z ERROR jd_client::lib::upstream_sv2::upstream: Failed to connect to Upstream role at 127.0.0.1:44254, retrying in 5s: Connection refused (os error 61)
2024-08-02T15:35:06.361785Z ERROR jd_client::lib::upstream_sv2::upstream: Failed to connect to Upstream role at 127.0.0.1:44254, retrying in 5s: Connection refused (os error 61)
```

am I missing something?

GitGab19 commented on roles/translator/src/main.rs:
```diff
@@ -77,6 +81,131 @@ async fn main() {
+    tx_status: async_channel::Sender<Status<'static>>,
+    task_collector: Arc<Mutex<Vec<(AbortHandle, String)>>>,
+) {
+    let proxy_config = match process_cli_args() {
+        Ok(p) => p,
+        Err(e) => panic!("failed to load config: {}", e),
+    };
```
Why do you need this here?
Where was it before your changes in this PR?

lorbax replied:
I dont need that there. I changed it passing to the start function

GitGab19 reviewed (commented):
LGTM. 
Since it seems not reliable (without the patch you mentioned us some days ago @lorbax), and we decided to move those integration tests in another place, I would just remove the MG test and the relative configs from this PR.
Tomorrow I will properly test it and review it again.

lorbax commented:
> onnection refused (os error 61)

this is the first time that I see this error. Usually the channel ID is fixed for the entire duration of the connection and starts from 1, but also in the MG test the message `SubmitShareError` has channel id is 1.

jbesraa commented on roles/translator/src/main.rs:
```diff
@@ -77,6 +81,129 @@ async fn main() {
+            State::UpstreamShutdown(err) => {
+                error!("SHUTDOWN from: {}", err);
+
+                // wait a random amount of time between 0 and 3000ms
```
Why do you need to introduce rand as a dep for this? just go for the 3000ms ?

plebhash replied:
as the comment states:

> if all the downstreams try to reconnect at the same time, the upstream may fail

if I understand correctly, this is trying to avoid a situation where multiple tProxies attempt to reconnect to the same Upstream at the same time, which could potentially cause it to fail

so it's attempting to introduce a randomness to this process, so that different tProxies will try to reconnect at different moments

I agree with the overall sentiment that this approach feels a bit hacky.

@lorbax do you know for a fact that the upstream will fail under these circumstances, or is it still just speculation?

if we're not sure this will happen, I'd say we should avoid solutions to imaginary problems

if we eventually see this happening, we can try to come up with a more sustainable solution

Fi3 replied:
this is best practice, when you have several client that can disconnect and then reconnect automatically you have to add a little randomness to the process. Is not hacky is the standard way to do this kind of thing

plebhash replied:
what if an attacker deploys multiple malicious tProxies and intentionally resets all the connections at the same time?

Fi3 replied:
is there to prevent to ddos yourself or the pool that you are using

jbesraa replied:
@Fi3 could you please explain further? I dont understand why this is needed

Fi3 replied:
you have 1M tproxy connected to a pool. The pool have to close the connection, all the tproxy try to reconnect in the same exact moment. Not good

plebhash replied:
I'm not strictly opposed to having this here, I just don't think there any substantial benefit

sure, we are making sure SRI tProxy is a "benign" client that intentionally avoids DDoSing the upstream connection

but in production scenarios, this is not really protecting upstream from malicious users

as discussed in the call, the only sustainable solution for this is load balancers and similar techniques to be deployed on the server infrastructure

Fi3 replied:
again this is industry best practice everyone does it. And there is a reason. This have nothing to do with protecting upstream from malicious users.  https://www.google.com/search?client=firefox-b-d&q=random+delay+backoff

Fi3 replied:
when you have an automatic reconnection or you retry to reconnect with exponetial backoff you add a random delay to avoid have all the client reconnect in the same instant.

Fi3 replied:
One thing is not ddos yourself another thing is protecting against a ddos

pavlenex replied:
@Fi3 Were this concerns addressed in https://github.com/stratum-mining/stratum/pull/1001#issuecomment-2270026282 it seems @lorbax agrees with the proposal, but it's not yet been committed?

Fi3 replied:
I think so. I did not reviewed it, I was only explaining why we should use randomized delay when we reconnect.

pavlenex replied:
@lorbax can you commit what @plebhash suggested if it addresses the concern? cc @plebhash 

Then we may be able to finally push this PR through and start testing it, and if it works https://github.com/stratum-mining/stratum/pull/1001#discussion_r1704210248 can be done in a follow up.

jbesraa commented on roles/translator/src/main.rs:
```diff
@@ -77,6 +81,129 @@ async fn main() {
+}
+
+#[allow(clippy::too_many_arguments)]
+async fn start<'a>(
```
ummm ?? could you please why all of these inputs?  how are you expecting other users to call this function?

plebhash replied:
I agree with the criticism, but I wonder if this couldn't be addressed in #1066 , since there's already some refactoring going on there?

lorbax replied:
> ummm ?? could you please why all of these inputs? how are you expecting other users to call this function?

I do not expect other users to call this function.

jbesraa replied:
This is a `start` function to the `translator` role, how are u not expecting other users to call it?
I mean, clippy is already screaming and you decide to disable it, I dont think this is the best approach

jbesraa reviewed (changes requested):
nack. I dont think the translator should look like that. the `start` function is far from anything we can work with. 
please look here https://github.com/stratum-mining/stratum/blob/fbb846450a3053f7e57c9bdf90d407d4e10449da/roles/translator/src/lib/mod.rs I think this should be more or less the approach.
If you need to persist data, then we should add write/read to (disk?) abilities and save the data and recover it when we need it later.

plebhash reviewed (changes requested):
I approved this earlier but then I noticed two MG tests are failing deterministically, so unfortunately I have to revert my approval.

Basically, now we're making tProxy attempt to reconnect to `Upstream` every time there's a `State::UpstreamShutdown`, regardless of what caused it.

## `pool-sri-test-close-channel`

To be honest, I'm not sure what this MG tries to achieve in the first place.

There's nothing [in the specs](https://stratumprotocol.org/specification/05-Mining-Protocol/#539-closechannel-client---server-server---client) stating that a proxy must close the entire connection when it receives a `CloseChannel` from `Upstream`.

This feels like a naive test. I would lean towards completely removing this test from the repo (ideally on a separate PR, to be merged before this one), but it would be good to get feedback from others.

## `translation-proxy-broke-pool`

This MG test asserts that the tProxy closes the connection when it receives a bad extranonce from `Upstream`.

That feels like a reasonable behavior, but it kind of conflicts with the current approach of this PR, because a bad extrononce also triggers a `State::UpstreamShutdown`, which now causes tProxy to attempt to reconnect. 

So if `Upstream` sends a bad extranonce, should tProxy really keep trying to reconnect forever?

I'm not sure how to proceed here.

---

@GitGab19 @lorbax please let me know your thoughts.

plebhash commented:
after further investigating the MG tests described above, I came up with a solution: https://github.com/plebhash/stratum/commit/e63c249bec12081115b7a41fcc999d5b97c5af99

basically, tProxy should try to reconnect if the error coming from `Upstream` is `Error::ChannelErrorReceiver(_)` or `Error::TokioChannelErrorRecv(_)`, which indicate an abrupt interruption on the channel managed by `network_helpers_sv2`

with this approach, both MG tests listed above are passing

I still think that `pool-sri-test-close-channel` is a bit weird and deserves a deeper investigation, but we can leave that for https://github.com/stratum-mining/stratum/issues/1081

@lorbax please have a look at the solution on the suggested commit... if you agree, feel free to just port it to your branch and squash it into the single commit of this PR, no need to worry about credit/authorship

lorbax commented:
> after further investigating the MG tests described above, I came up with a solution: [plebhash@e63c249](https://github.com/plebhash/stratum/commit/e63c249bec12081115b7a41fcc999d5b97c5af99)
> 
> basically, tProxy should only try to reconnect if the error coming from `Upstream` is `Error::ChannelErrorReceiver(_)` or `Error::TokioChannelErrorRecv(_)`, which indicates an abrupt interruption on the channel managed by `network_helpers_sv2`
> 
> with this approach, both MG tests listed above are passing
> 
> I still think that `pool-sri-test-close-channel` is a bit weird and deserves a deeper investigation, but we can leave that for #1081
> 
> @lorbax please have a look at the solution on the suggested commit... if you agree, feel free to just port it to your branch and squash it into the single commit of this PR, no need to worry about credit/authorship

looks good

lorbax commented:
> after further investigating the MG tests described above, I came up with a solution: [plebhash@e63c249](https://github.com/plebhash/stratum/commit/e63c249bec12081115b7a41fcc999d5b97c5af99)
> 
> basically, tProxy should only try to reconnect if the error coming from `Upstream` is `Error::ChannelErrorReceiver(_)` or `Error::TokioChannelErrorRecv(_)`, which indicates an abrupt interruption on the channel managed by `network_helpers_sv2`
> 
> with this approach, both MG tests listed above are passing
> 
> I still think that `pool-sri-test-close-channel` is a bit weird and deserves a deeper investigation, but we can leave that for #1081
> 
> @lorbax please have a look at the solution on the suggested commit... if you agree, feel free to just port it to your branch and squash it into the single commit of this PR, no need to worry about credit/authorship

Hi!
I had a closer look to your proposal. When the Translator gets disconnected, it receives 
```rust
translator_sv2::error::Error::ChannelErrorReceiver
```
but you suggests to send `UpstreamTryReconnect`. Did you have an actual case in which the jdc disconnects the Translator and you get 
```rust
translator_sv2::error::Error::TokioChannelErrorRecv
```
?
why it is not enough the following?
```rust
        Sender::Upstream(tx) => {
            tx.send(Status {
            match e {
                state: State::UpstreamShutdown(e),
                Error::ChannelErrorReceiver(_) => {
            })
                    tx.send(Status {
            .await
                        state: State::UpstreamTryReconnect(e),
            .unwrap_or(());
                    })
                        .await
                        .unwrap_or(());
                },
                _ => {
                    tx.send(Status {
                        state: State::UpstreamShutdown(e),
                    })
                        .await
                        .unwrap_or(());
                },
            }
        }
    }

plebhash commented:
@lorbax `network_helpers_sv2` offers two flavors of channels: `async_std` and `tokio`: https://github.com/stratum-mining/stratum/tree/dev/roles/roles-utils/network-helpers/src

so I just tried to cover for both flavors

but feel free to only do `ChannelErrorReceiver` if you think that's overkill

lorbax commented:
> @lorbax `network_helpers_sv2` offers two flavors of channels: `async_std` and `tokio`: https://github.com/stratum-mining/stratum/tree/dev/roles/roles-utils/network-helpers/src
> 
> so I just tried to cover for both flavors
> 
> but feel free to only do `ChannelErrorReceiver` if you think that's overkill

my idea is to introduce changes only when strictly needed in order to avoid unexpected behavior. What do you think about?
again, I stress also that we are using two types of channels, from tokio and from async_std. It seems a red flag to me, what do you think also about this?

plebhash commented:
> > @lorbax `network_helpers_sv2` offers two flavors of channels: `async_std` and `tokio`: https://github.com/stratum-mining/stratum/tree/dev/roles/roles-utils/network-helpers/src
> > so I just tried to cover for both flavors
> > but feel free to only do `ChannelErrorReceiver` if you think that's overkill
> 
> my idea is to introduce changes only when strictly needed in order to avoid unexpected behavior. What do you think about? again, I stress also that we are using two types of channels, from tokio and from async_std. It seems a red flag to me, what do you think also about this?

yes those are very good points and I agree it's kind of a red flag... I'm not sure what are the historical motivations for having two different flavors of channels, but I think that when we start making [deeper refactoring into the application layer](https://github.com/stratum-mining/stratum/discussions/1069#discussioncomment-10290635) we should discuss the possibility of getting rid of some of the two kinds of channels in favor of standardization

but that's a discussion for the future

for now, on the context of this PR, I agree we should move forward only covering `ChannelErrorReceiver` on the `match` arms and just ignore the `TokioChannelErrorRecv`, since `tokio` channels are not even enabled on the `network_helpers_sv2` feature flags:

https://github.com/stratum-mining/stratum/blob/2541703f2f5371052110f740f7eab9409a10d0f1/roles/translator/Cargo.toml#L22

GitGab19 commented on roles/translator/src/main.rs:
```diff
@@ -54,22 +54,122 @@ async fn main() {
+
+                // wait a random amount of time between 0 and 3000ms
+                // if all the downstreams try to reconnect at the same time, the upstream may fail
+                tokio::time::sleep(Duration::from_millis(1000)).await;
```
Isn't this line redundant?
You already wait a random time below

GitGab19 commented on roles/translator/src/main.rs:
```diff
@@ -54,22 +54,122 @@ async fn main() {
+                let task_collector_aborting = task_collector_.clone();
+                kill_tasks(task_collector_aborting.clone());
+
+                error!("Trying recconnecting to upstream");
```
Shouldn't be a warning?
Also, there's a typo: it should be "Trying reconnecting to upstream"

GitGab19 reviewed (commented):
Tested twice and it works as expected, good work @lorbax !
Just a few minor nits and it's ready to go

lorbax commented:
> > > @lorbax `network_helpers_sv2` offers two flavors of channels: `async_std` and `tokio`: https://github.com/stratum-mining/stratum/tree/dev/roles/roles-utils/network-helpers/src
> > > so I just tried to cover for both flavors
> > > but feel free to only do `ChannelErrorReceiver` if you think that's overkill
> > 
> > 
> > my idea is to introduce changes only when strictly needed in order to avoid unexpected behavior. What do you think about? again, I stress also that we are using two types of channels, from tokio and from async_std. It seems a red flag to me, what do you think also about this?
> 
> yes those are very good points and I agree it's kind of a red flag... I'm not sure what are the historical motivations for having two different flavors of channels, but I think that when we start making [deeper refactoring into the application layer](https://github.com/stratum-mining/stratum/discussions/1069#discussioncomment-10290635) we should discuss the possibility of getting rid of some of the two kinds of channels in favor of standardization
> 
> but that's a discussion for the future
> 
> for now, on the context of this PR, I agree we should move forward only covering `ChannelErrorReceiver` on the `match` arms and just ignore the `TokioChannelErrorRecv`, since `tokio` channels are not even enabled on the `network_helpers_sv2` feature flags:
> 
> https://github.com/stratum-mining/stratum/blob/2541703f2f5371052110f740f7eab9409a10d0f1/roles/translator/Cargo.toml#L22

with channel I mean channels that allows a task to send data to another task, like mpsc channels.
In the translator we are using async_std channels
https://docs.rs/async-std/latest/async_std/channel/fn.bounded.html
but the error variant also take in consideration the mpmc channel in tokio
https://docs.rs/tokio/latest/tokio/sync/broadcast/index.html

These are not things that are exported by network-helpers (or at least it appears to me like this). Perhaps is a good idea to make it uniform and substitute the usage of  `async_std::channel` with `tokio::sync::mpsc::channel`. But in this case we must recall to change it also in `send_status` function

lorbax commented:
> > > > @lorbax `network_helpers_sv2` offers two flavors of channels: `async_std` and `tokio`: https://github.com/stratum-mining/stratum/tree/dev/roles/roles-utils/network-helpers/src
> > > > so I just tried to cover for both flavors
> > > > but feel free to only do `ChannelErrorReceiver` if you think that's overkill
> > > 
> > > 
> > > my idea is to introduce changes only when strictly needed in order to avoid unexpected behavior. What do you think about? again, I stress also that we are using two types of channels, from tokio and from async_std. It seems a red flag to me, what do you think also about this?
> > 
> > 
> > yes those are very good points and I agree it's kind of a red flag... I'm not sure what are the historical motivations for having two different flavors of channels, but I think that when we start making [deeper refactoring into the application layer](https://github.com/stratum-mining/stratum/discussions/1069#discussioncomment-10290635) we should discuss the possibility of getting rid of some of the two kinds of channels in favor of standardization
> > but that's a discussion for the future
> > for now, on the context of this PR, I agree we should move forward only covering `ChannelErrorReceiver` on the `match` arms and just ignore the `TokioChannelErrorRecv`, since `tokio` channels are not even enabled on the `network_helpers_sv2` feature flags:
> > https://github.com/stratum-mining/stratum/blob/2541703f2f5371052110f740f7eab9409a10d0f1/roles/translator/Cargo.toml#L22
> 
> with channel I mean channels that allows a task to send data to another task, like mpsc channels. In the translator we are using async_std channels https://docs.rs/async-std/latest/async_std/channel/fn.bounded.html but the error variant also take in consideration the mpmc channel in tokio https://docs.rs/tokio/latest/tokio/sync/broadcast/index.html
> 
> These are not things that are exported by network-helpers (or at least it appears to me like this). Perhaps is a good idea to make it uniform and substitute the usage of `async_std::channel` with `tokio::sync::mpsc::channel`. But in this case we must recall to change it also in `send_status` function

jbesraa commented:
@GitGab19 lets not merge unless we have all CI checks in the green please
